src/Microsoft.Data/PrimitiveColumn.Sort.cs

```diff
@@ -39,16 +42,33 @@ private void GetSortIndices(IComparer<T> comparer, out PrimitiveColumn<long> col
                 List<int> nonNullSortIndices = new List<int>();
                 for (int i = 0; i < sortIndices.Length; i++)
                 {
-                    if (IsValid(sortIndices[i] + b * ReadOnlyDataFrameBuffer<T>.MaxCapacity))
+                    if (_columnContainer.IsValid(ref nullBitMapSpan, sortIndices[i]))
                         nonNullSortIndices.Add(sortIndices[i]);
+
                 }
                 bufferSortIndices.Add(nonNullSortIndices);
             }
             // Simple merge sort to build the full column's sort indices
             ValueTuple<T, int> GetFirstNonNullValueAndBufferIndexStartingAtIndex(int bufferIndex, int startIndex)
             {
-                T value = _columnContainer.Buffers[bufferIndex][bufferSortIndices[bufferIndex][startIndex]];
-                long rowIndex = bufferSortIndices[bufferIndex][startIndex] + bufferIndex * ReadOnlyDataFrameBuffer<T>.MaxCapacity;
+                int index = bufferSortIndices[bufferIndex][startIndex];
+                T value;
+                ReadOnlyMemory<byte> buffer = _columnContainer.Buffers[bufferIndex].ReadOnlyMemory;
+                ReadOnlyMemory<T> typedBuffer = Unsafe.As<ReadOnlyMemory<byte>, ReadOnlyMemory<T>>(ref buffer);
+                if (!typedBuffer.IsEmpty)
+                {
+                    bool isArray = MemoryMarshal.TryGetArray(typedBuffer, out ArraySegment<T> array);
+                    if (isArray)
+                    {
+                        value = array.Array[index];
```

Comment:
You need to use the ArraySegment.Offset here.  The ArraySegment might point to somewhere inside the array, and not at the beginning.

Reply:
Hmm, I don't understand, array.Array will return the original `Array` right? Shouldn't it be safe to then index into the original `Array`?

Reply:
```C#
MemoryMarshal.TryGetArray(typedBuffer, out ArraySegment<T> array);
```

`TryGetArray` and naming `ArraySegment<T>` as `array` are confusing you. `ArraySegment` is a section of an array. `typedBuffer` is a `ReadOnlyMemory`. A `Memory` instance may be pointing into the middle of an array. So say the whole backing array might look like:

```
[5, 4, 3, 2, 1]
```

`typedBuffer` might be pointing at just the `4, 3, 2` part of the total backing array.

That's why this method returns an `ArraySegment` - a section of the array. In the example above, you'd get the original `[5, 4, 3, 2 , 1]` array back, but the `ArraySegment.Offset` will be `1`, meaning the segment of the array that the buffer is pointing to starts at element `1` of the array.

In your code, `index` is the index into `typedBuffer`. And `typedBuffer` starts at `ArraySegment.Offset` into `ArraySegment.Array`. So you need to add `index` and `ArraySegment.Offset` together to get the total index into the backing array.

Reply:
Ah, now I see why you said that. So, in this case, `typedBuffer` would always be the full buffer, never a slice. So `ArraySegment.Offset` would always be 0. But, I get the point. I'll change it to `ArraySegment.Offset + index` just in case `typedBuffer` stops being the full buffer in the future.

Reply:
> in this case, typedBuffer would always be the full buffer, never a slice.

How are you guaranteeing that? What if the data came from Apache Arrow?  There, we might have taken a giant `byte[]` and sliced it all up to form the individual Arrow columns.

Reply:
Just a general note of caution (as I am passing by). Try to avoid the use of "unsafe" types/classes like `MemoryMarshal` (basically anything in the `S.R.InteropServices` namespace) and `S.R.CS.Unsafe`. I would default to use the safe APIs and only use the unsafe/advance APIs in limited/targeted scenarios where it is clear that the perf wins are noticeable and worth the maintenance burden (and potential correctness risk). That said, I haven't looked deeply into its usage in the context of this PR/feature, so treat this note as just a heads up to scrutinize usage of such APIs more closely during code review (I got the sense that a bunch of type-safety/etc. was being forgone here).

Reply:
> (I got the sense that a bunch of type-safety/etc. was being forgone here)

In general, the Arrow format forces a bunch of type-safety to be forgone. At its heart, Arrow takes a collection of bytes, and reinterprets them as different types depending on the type information stored in the Arrow metadata. So while we have a `Memory<byte>` underneath, we need to reinterpret it as a `Memory<int|float|etc>` later in order to read/write integers, floats, etc from the data.